sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java

```diff
@@ -372,14 +385,57 @@ public BoundedWindow window() {
-   * {@link DoFn} must have a method with this annotation satisfying
-   * the following constraints in order for it to be executable:
+   * Annotation for the method to use for processing elements. A subclass of {@link DoFn} must have
+   * a method with this annotation.
+   *
+   * <p>The signature of this method must satisfy the following constraints:
+   *
    * <ul>
-   *   <li>It must have at least one argument.
-   *   <li>Its first argument must be a {@link DoFn.ProcessContext}.
-   *   <li>Its remaining argument, if any, must be {@link BoundedWindow}.
+   * <li>Its first argument must be a {@link DoFn.ProcessContext}.
+   * <li>Its remaining arguments, if any, must be either {@link BoundedWindow} or a subtype of
+   *     {@link RestrictionTracker}. Each of these must appear at most once.
+   * <li>The {@link BoundedWindow} argument corresponds to the window of the current element. If
+   *     absent, a runner may perform additional optimizations.
+   * <li>If the {@link RestrictionTracker} argument is present, this {@link DoFn} is treated as <a
+   *     href="https://s.apache.org/splittable-do-fn">splittable</a> and there are additional
+   *     constraints on related methods {@link GetInitialRestriction}, {@link SplitRestriction},
+   *     {@link NewTracker}, {@link GetRestrictionCoder}, as well as the class-level annotations
+   *     {@link Bounded} and {@link Unbounded}.
+   * <li>If this {@link DoFn} is splittable, the {@link ProcessElement} method <i>may</i> return a
+   *     {@link ProcessContinuation} to indicate whether there is more work to be done for the
+   *     current element.
    * </ul>
+   *
+   * <h2>Splittable DoFn's</h2>
+   *
+   * A {@link DoFn} is <i>splittable</i> if its {@link ProcessElement} method has a parameter whose
+   * type is a subtype of {@link RestrictionTracker}. This is an advanced feature and an
+   * overwhelming majority of users will never need to write a splittable {@link DoFn}. Right now
+   * the implementation of this feature is in progress and it's not ready for any usage.
+   *
+   * <p>See <a href="https://s.apache.org/splittable-do-fn">the proposal</a> for an overview of the
+   * involved concepts (<i>splittable DoFn</i>, <i>restriction</i>, <i>restriction tracker</i>).
+   *
+   * <p>If a {@link DoFn} is splittable, the following constraints must be respected:
+   *
+   * <ul>
+   * <li>It <i>must</i> define a {@link GetInitialRestriction} method.
+   * <li>It <i>may</i> define a {@link SplitRestriction} method.
+   * <li>It <i>must</i> define a {@link NewTracker} method returning the same type as the type of
+   *     the {@link RestrictionTracker} argument of {@link ProcessElement}, which in turn must be a
+   *     subtype of {@code RestrictionTracker<R>} where {@code R} is the restriction type returned
+   *     by {@link GetInitialRestriction}.
+   * <li>It <i>must</i> define a {@link GetRestrictionCoder} method.
```

Comment:
Couldn't the RestrictionCoder can't be part of the RestrictionTracker? If RestrictionTracker had a `getRestrictionCoder` method that received the input/output coders, it could encapsulate the logic to encode a restriction of type `R`, correct? That would reduce the burden on each `DoFn` using that `RestrictionTracker`.